homeassistant/components/opentherm_gw/binary_sensor.py

```diff
@@ -6,43 +6,66 @@
 from homeassistant.core import callback
 from homeassistant.helpers.dispatcher import async_dispatcher_connect
 from homeassistant.helpers.entity import async_generate_entity_id
+from homeassistant.helpers.entity_registry import async_get_registry
 
 from . import DOMAIN
-from .const import BINARY_SENSOR_INFO, DATA_GATEWAYS, DATA_OPENTHERM_GW
+from .const import (
+    BINARY_SENSOR_INFO,
+    DATA_GATEWAYS,
+    DATA_OPENTHERM_GW,
+    TRANSLATE_SOURCE,
+)
 
 _LOGGER = logging.getLogger(__name__)
 
 
 async def async_setup_entry(hass, config_entry, async_add_entities):
     """Set up the OpenTherm Gateway binary sensors."""
     sensors = []
+    gw_dev = hass.data[DATA_OPENTHERM_GW][DATA_GATEWAYS][config_entry.data[CONF_ID]]
+    ent_reg = await async_get_registry(hass)
     for var, info in BINARY_SENSOR_INFO.items():
+        old_style_entity_id = async_generate_entity_id(
+            ENTITY_ID_FORMAT, f"{var}_{gw_dev.gw_id}", hass=gw_dev.hass
+        )
+        old_ent = ent_reg.async_get(old_style_entity_id)
+        if old_ent and old_ent.config_entry_id == config_entry.entry_id:
+            ent_reg.async_remove(old_style_entity_id)
```

Comment:
Is it impossible to migrate the old entries to be part of the new entry structure?

Reply:
Currently the old entries carry information mashed together from all sources (thermostat, boiler and gateway itself) without a clear indication where the information is coming from. Some entities even carried information from 2 sources at the same time, which was causing #39752 if the sources disagreed.

If we were to integrate these old entities in the new system, the clearest way to do it in my opinion would be to assign all of the old entries to a single source. This would still change the source of some of the sensors and thus it would still be a breaking change. All automations/scripts/etc. depending on these entities would still work, but may have changed their behavior very subtly from previous versions.

If we remove all old entities we force people to reconsider their automations/scripts and adapt them to the new style entities. This way it is always clear where the information in the sensor is coming from and we don't change existing functionality in any entities that depend on our sensors and binary sensors without breaking them.

In my opinion the first scenario (subtle change of behavior without a hard break) is not desirable. The latter option (hard break forcing people to review) has my personal preference. It also keeps things simple in the entity classes as we don't need to take exceptions without a `source` into account.
Please also take note of the fact that we are only breaking entities that are disabled by default. Only people who actively enabled any sensors or binary sensors from the opentherm_gw integration are affected. The climate entity - which is the main "interface" - is unaffected.

So yes, technically it would be possible to migrate the old entries to the new structure. I just have my doubts about whether it would be the best option in this case.

Reply:
Would it be possible to keep the old entities with a deprecation message and and also add the new entities?

Reply:
That's a great idea! Do we have any deprecation facilities built into HA or would a log message and possibly a notification suffice?
I plan to implement this for enabled entities only and to just delete the disabled ones. This way only people that are actually using the old entities will see a deprecation message.

Reply:
Just log a warning. There's no special tools for deprecating entities.

Sounds good!

Reply:
Added in 21bb048. I tried to match the behavior of the deprecated entities as much as possible to the behavior of `pyotgw==0.6b1` in the process.